### ISSUE TITLE
Help needed: Wrap siblings to one parent wrapper

### ISSUE THREAD

**Issue by ManasMadrecha** [jake-low/remark-sectionize#9]
Hi, first of all, thanks for this amazing plugin.

This is not an issue, per se. As this plugin wraps elements into a wrapper section, I thought this is a right place to ask a question on creating similar plugin.

#### 1

I want to wrap this:

```md
> this is a blockquote
<figcaption></figcaption>
```
 into this:

```md
<figure>
> this is a blockquote
<figcaption></figcaption>
</figure>
```

#### 2

```md
term

: definition
: definition
```
 into this:

```md
<div>

term

: definition
: definition

</div>
```

#### 3

```md
term

: definition
: definition


term2

: definition2
: definition2a
```
 into this:

```md
<div>

term

: definition
: definition

</div>

<div>

term2

: definition2
: definition2a

</div>
```

Currently, https://github.com/Symbitic/remark-plugins/blob/master/packages/remark-deflist/src/index.js this plugin converts term and definition into proper HTML.

```md
<dl>

<dt></dt>
<dd></dd>
<dd></dd>

</dl>
```

and

```md
<dl>

<dt></dt>
<dd></dd>
<dd></dd>

<dt></dt>
<dd></dd>
<dd></dd>
</dl>
```

So, basically I want to have a wrapper set for some sibling elements. Can you kindly make one plugin for this or guide as to how to proceed?

**Comment by jake-low** [comment]
Hi Manas,

It should be fairly straightforward for you to modify the code of this plugin to suit your needs. The plugin is MIT-licensed so you can use the code pretty much however you want, as long as you preserve the LICENSE file in your fork.

I'll offer some guidance here to get you started.

All the code lives in one short file, [index.js](https://github.com/jake-low/remark-sectionize/blob/master/index.js). The main export is a function called `plugin()`. Remark plugins are functions that take an `options` object (in this case there are no options) and return another function that does the actual work. In this case that function is called `transform`.

The `transform` function takes a tree of Remark nodes and mutates it. In this case, it modifies the tree to wrap each heading and the following content in a section. It does this iteratively (see the for loop on line 13), one heading depth at a time. For each possible depth, it "visits" all heading nodes of that depth and calls `sectionize` on them.

The `sectionize` function which does the actual wrapping takes a node in the tree (which we know to be a heading node, because of the filter function we passed into `visit` above) and wraps it and all the content after it in a section. It does this getting the parent of the heading node (line 24) and then finding the next child of that parent node after the current heading node which satisfies the `isEnd` condition (lines 27-28). Then it slices the range of nodes between the heading node and the end node out of the parent's children (lines 33-36), and replaces it with a new section node that contains those children instead (lines 38-47).

To adapt this code to your purposes, you can probably remove the for loop in `transform`, and just call `visit` once. You'd pass a filter function (the second argument to `visit`) that selected the type of node which you want to start wrapping at (e.g. `node.type === 'blockquote'`). Then `sectionize` (which you'd probably rename to `wrap` or something) would do similar logic as it currently does, but with a different `isEnd` function. Perhaps you'd check if the end node was a figcaption, and then adjust the `endIndex` by one to include the end node itself. Finally, you'd splice in a new wrapper node, only instead of `type: 'section'` you'd make it whatever type you want the wrapper to be. For a `figure`, you'd need to be sure to add an `hName` property so that rehype knows what HTML element to convert the node to.

Hope that helps!
